doc/man/man4/nit_if.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/nit_if.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the section entitled "GNU Free
.\" Documentation License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaning to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF
.\" SUCH CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH NIT_IF 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Modules"
.\"
.\"
.SH NAME
.B nit_if
\- \fISTREAMS\fP Network Interface Tap (SNIT) device driver
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <sys/types.h>
.br
.B #include <sys/stropts.h>
.br
.B #include <sys/stream.h>
.br
.B #include <sys/ioctl.h>
.br
.B #include <net/nit_if.h>
.HP 8
.BI "int " fd ;
.PD 0
.HP
.BI "int open(" \*(lq\fR/dev/streams/nit/\fP\fIxxxx\fP\*(rq ", " \fR0\fP );
.PD
.\"
.\"
.SH DESCRIPTION
.PP
.B nit
is a
.I STREAMS
pseudo-device driver that provides
.I STREAMS
access to network interfaces.  It is designed to be used in conjunction with
the other components of NIT,
.BR nit (4),
but can be used by itself as a raw
.I STREAMS
network interface.
.PP
.B nit
is an exclusive-open device that is intended to be opened indirectly through
the cloning device,
.BR /dev/streams/nit .
A suitable device driver path to open is
.BI /dev/streams/nit/ xxxx,
where
.I xxxx
is the name of a NIT supporting device driver.
Before the stream resulting from opening an instance of
.B nit
may be used to read or write packets, it must first be bound to a specific
network interface, using the
.B NIOCSBIND
input-output control command described below.
.\"
.SS Read-side Behaviour
.PP
.B nit
copies leading prefixes of selected packets from its associated network
interface and passes them up the stream.  If the
.B NI_PROMISC
flag is set, it passes along all packets; otherwise, it passes along only
packets addressed to the underlying interface.
.PP
The amount of data copied from a given packet depends on the current snapshot
length, which is set with the
.B NIOCSSNAP
input-output control command described below.
.PP
Before passing each packet prefix upward,
.B nit
optionally prepends one or more headers, as controlled by the state of the
flag bits set with the
.B NIOCSFLAGS
input-output control command.  The driver collects headers into
.BR M_PROTO (9)
message blocks, with the headers guaranteed to be completely contained in a
single message block, whereas the packet itself goes into one or more
.BR M_DATA (9)
message blocks.
.\"
.SS Write-side Behaviour
.PP
.B nit
accepts packets from the module above it in the stream and relays them to the
associated network interface for transmission.  Packets must be formatted
with the destination address in a leading
.BR M_PROTO (9)
message block, followed by the packet itself, complete with link-level
header, in a sequence of
.BR M_DATA (9)
message blocks.  The destination address must be expressed as a
.B struct sockaddr
whose
.I sa_family
field is
.B AF_UNSPEC
and whose
.I sa_data
field is a copy of the link-level header.  (See
.I <sys/socket.h>
for the definition of this structure.)  If the packet does not conform to
this format, an
.BR M_ERROR (9)
message with
.RB [ EINVAL ]
will be sent upstream.
.PP
.B nit
processes
.BR M_IOCTL (9)
messages as described below.  Upon receiving an
.BR M_FLUSH (9)
message specifying that the write queue be flushed,
.B nit
does so and transfers the message to the read side of the stream.  It
discards all other messages.
.\"
.\"
.SH IOCTLS
.PP
.B nit
response to the following input-output control commands, as defined in
.IR <net/nit_if.h> .
It generates an
.BR M_IOCNAK (9)
message for all others, returning this message to the invoker along the
read side of the stream.
.PP
.TP
.BR SIOCGIFADDR ", " SIOCADDMULTI ", " SIOCDELMULTI
.B nit
passes these input-output control commands on to the underlying interface's
driver and returns it response in a
.B struct ifreq
instance, as defined in
.IR <net/if.h> .
(See the description of this input-output control command in
.BR if (4)
for more details.)
.TP
.B NIOCBIND
This input-output control command will attach the stream represented by its
first argument to the network interface designated by its third argument,
which should be a pointer to an
.B ifreq
structure whose
.I ifr_name
field names the desired interface.  See
.I <net/if.h>
for the definition of this structure.
.TP
.B NIOCSSNAP
Set the current snapshot length to the value given in the
.B u_long
pointed to by the input-output control command's final argument.
.B nit
interprets a snapshot length of zero as meaning infinity, so that it will
copy all selected packets in their entirety.  It constrains positive snapshot
lengths to be at least the length of an Ethernet header, so that it will pass
at least the link-level header of all selected packets to its upstream
neighbour.
.TP
.B NIOCGSNAP
Returns the current snapshot length for this device instance in the
.B u_long
pointed to by the input-output control command's final argument.
.TP
.B NIOCSFLAGS
.B nit
recognizes the following flag bits, which must be given in the
.B u_long
pointed to by the input-output control command's final argument.  This set
may be augmented in future releases.  All but the
.I NI_PROMISC
bit control the addition of headers that precede the packet body.  These
headers appear in the order given below, with the last-mentioned enabled
header adjacent to the packet body.
.RS
.TP
.I NI_PROMISC
Requests that the underlying interface be set into promiscuous mode and that
all packets that the interface receives be passed up through the stream.
.B nit
only honours this bit for the super-user.
.TP
.I NI_TIMESTAMP
Prepend to each selected packet a header containing the packet arrival time
expressed as a
.BR "struct timeval" .
.TP
.I NI_DROPS
Prepend to each selected packet a header containing the cumulative number of
packets that this instance of
.B nit
has dropped because of flow control requirements or resource exhaustion.  The
header value is expressed as a
.BR u_long .
Note: it accounts only for events occurring within
.BR nit ,
and does not count packets dropped at the network interface level or by
upstream modules.
.TP
.I NI_LEN
Prepend to each selected packet a header containing the packet;s original
length (including link-level header), as it was before being trimmed to the
snapshot length.  The header value is expressed as
.BR u_long .
.RE
.TP
.B NIOCGFLAGS
Returns the current state of the flag bits for this device instance in the
.B u_long
pointed to by the input-output control command's final argument.
.\"
.\"
.SH FILES
.PP
.TP
.IR /dev/nit " or " /dev/streams/clone/nit
The clone device instance referring to the
.B nit
device.
.TP
.I <net/nit_if.h>
Header file containing definitions for the input-output control commands
and packet headers described above.
.\"
.\"
.SH IMPLEMENTATION
.PP
Note that
.B nit
is impemented using the same driver as is provided for
.BR dlpi (4);
it just responds in the same manner as would the
.B nit
driver when opened as
.RB \*(lq /dev/nit \*(rq.
.\"
.\"
.SH CAVEATS
.PP
.\"
.\"
.SH "SEE ALSO"
.PP
.BR clone (4),
.BR nit (4),
.BR nit_buf (4),
.BR nit_pf (4).
.\"
.\"
.SH COMPATIBILITY
.PP
.B nit_buf
is compatible with
.BR "SVR 4.1" ,
.[
svr41spg
.]
and descriptions for
.BR SunOS\(rg 4.x,
with the following portability considerations:
.IP \(em \w'\(em\ \ 'u
.IP \(em
.PP
Compatibility is tested using the
.BR test-snit (8)
test case executable and the
.I @PACKAGE_TITLE@
.B autotest
test suite.
.\"
.\"
.SH CONFORMANCE
.PP
Network interface taps and raw sockets and DLPI drivers have never been an
interesting subject for standards organizations such as IEEE and the
OpenGroup, and these modules have never been subjected to formal
standardization.
.\"
.\"
.SH HISTORY
.PP
.B snit
first appeared in
.IR "SunOS\fR\(rg\fP 4.x" .
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/drivers/nit_if.c
/*****************************************************************************

 @(#) File: src/drivers/nit_if.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/drivers/nit_if.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

/*
 * This is a nit driver that works like the clone driver.  It accepts registration from other
 * drivers and provides a /dev/streams/nit major device that has a /dev/streams/nit/[driver] minor
 * device for each driver that registers.  [driver] is the driver name.  When the device is opened,
 * instead of forwarding to (minor,0) as does the clone driver, it forwards to (minor,1) with the
 * CLONEOPEN flag set.  There is also a dlpi and bpf driver that does a similar thing.
 */

#ifdef NEED_LINUX_AUTOCONF_H
#include NEED_LINUX_AUTOCONF_H
#endif
#include <linux/version.h>
#include <linux/module.h>
#include <linux/init.h>
#include <linux/sched.h>

#include <sys/stream.h>
#include <sys/strconf.h>
#include <sys/strsubr.h>
#include <sys/ddi.h>

#include <net/nit_if.h>		/* extern verification */

#include "sys/config.h"

#define NIT_DESCRIP	"SVR 4.1 SNIT (NIT_IF) STREAMS Driver"
#define NIT_EXTRA	"Part of UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define NIT_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define NIT_REVISION	"OpenSS7 src/drivers/nit_if.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define NIT_DEVICE	"SVR 4.2 MP STREAMS NIT Driver"
#define NIT_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define NIT_LICENSE	"GPL"
#define NIT_BANNER	NIT_DESCRIP	"\n" \
			NIT_EXTRA	"\n" \
			NIT_COPYRIGHT	"\n" \
			NIT_REVISION	"\n" \
			NIT_DEVICE	"\n" \
			NIT_CONTACT	"\n"
#define NIT_SPLASH	NIT_DEVICE	" - " \
			NIT_REVISION	"\n"

#ifdef CONFIG_STREAMS_NIT_MODULE
MODULE_AUTHOR(NIT_CONTACT);
MODULE_DESCRIPTION(NIT_DESCRIP);
MODULE_SUPPORTED_DEVICE(NIT_DEVICE);
MODULE_LICENSE(NIT_LICENSE);
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif				/* MODULE_VERSION */
#endif				/* CONFIG_STREAMS_NIT_MODULE */

#ifndef CONFIG_STREAMS_NIT_NAME
#error "CONFIG_STREAMS_NIT_NAME must be defined."
#endif
#ifndef CONFIG_STREAMS_NIT_MODID
#error "CONFIG_STREAMS_NIT_MODID must be defined."
#endif
#ifndef CONFIG_STREAMS_NIT_MAJOR
#error "CONFIG_STREAMS_NIT_MAJOR must be defined."
#endif

#ifdef CONFIG_STREAMS_NIT_MODULE
modID_t nit_modid = CONFIG_STREAMS_NIT_MODID;

#ifndef module_param
MODULE_PARM(nit_modid, "h");
#else
module_param(nit_modid, ushort, 0444);
#endif
MODULE_PARM_DESC(nit_modid, "Module id number for NIT driver.");
#else				/* CONFIG_STREAMS_NIT_MODULE */
static modID_t nit_modid = CONFIG_STREAMS_NIT_MODID;
#endif				/* CONFIG_STREAMS_NIT_MODULE */

#ifdef CONFIG_STREAMS_NIT_MODULE
major_t major = CONFIG_STREAMS_NIT_MAJOR;

#ifndef module_param
MODULE_PARM(major, "h");
#else
module_param(major, uint, 0444);
#endif
MODULE_PARM_DESC(major, "Major device number for NIT driver.");
#else				/* CONFIG_STREAMS_NIT_MODULE */
static major_t major = CONFIG_STREAMS_NIT_MAJOR;
#endif				/* CONFIG_STREAMS_NIT_MODULE */

#ifdef MODULE
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-nit");
MODULE_ALIAS("streams-modid-" __stringify(CONFIG_STREAMS_NIT_MODID));
MODULE_ALIAS("streams-driver-nit");
MODULE_ALIAS("char-major-" __stringify(CONFIG_STREAMS_CLONE_MAJOR) "-" __stringify(CONFIG_STREAMS_NIT_MAJOR));
MODULE_ALIAS("/dev/nit");
//MODULE_ALIAS("devname:nit");
MODULE_ALIAS("streams-major-" __stringify(CONFIG_STREAMS_NIT_MAJOR));
MODULE_ALIAS("/dev/streams/nit");
MODULE_ALIAS("/dev/streams/nit/*");
#endif
#endif

static struct module_info nit_minfo = {
	.mi_idnum = CONFIG_STREAMS_NIT_MODID,
	.mi_idname = CONFIG_STREAMS_NIT_NAME,
	.mi_minpsz = STRMINPSZ,
	.mi_maxpsz = STRMAXPSZ,
	.mi_hiwat = STRHIGH,
	.mi_lowat = STRLOW,
};

static struct module_stat nit_rstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));
static struct module_stat nit_wstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));

static struct qinit nit_rinit = {
	.qi_putp = strrput,
	.qi_srvp = strrsrv,
	.qi_qopen = str_open,
	.qi_qclose = str_close,
	.qi_minfo = &nit_minfo,
	.qi_mstat = &nit_rstat,
};

static struct qinit nit_winit = {
	.qi_putp = strwput,
	.qi_srvp = strwsrv,
	.qi_minfo = &nit_minfo,
	.qi_mstat = &nit_wstat,
};

#ifdef CONFIG_STREAMS_NIT_MODULE
static
#endif
struct streamtab nitinfo = {
	.st_rdinit = &nit_rinit,
	.st_wrinit = &nit_winit,
};

/**
 *  nitopen: - open a nit special device
 *  @inode:	shadow special filesystem inode
 *  @file:	shadow special filesystem file pointer
 *
 *  nitopen() is called only from within the shadow special filesystem.  This can occur by
 *  chaining from the external filesystem (e.g. openining a character device with nit major) or by
 *  direct open of the inode within the mounted shadow special filesystem.  Either way, the inode
 *  number has our extended device numbering as a inode number and we chain the call within the
 *  shadow special filesystem.
 */
static int
nitopen(struct inode *inode, struct file *file)
{
	struct cdevsw *cdev;
	dev_t dev = inode->i_ino;

	if (file->private_data)
		/* Darn.  Somebody passed us a FIFO inode. */
		return (-EIO);

	if ((cdev = sdev_get(getminor(dev)))) {
		int err;

		dev = makedevice(cdev->d_modid, NIT_NODE);
		err = spec_open(file, cdev, dev, CLONEOPEN);
		sdev_put(cdev);
		return (err);
	}
	return (-ENOENT);
}

struct file_operations nit_ops ____cacheline_aligned = {
	.owner = THIS_MODULE,
	.open = nitopen,
};

/* 
 *  -------------------------------------------------------------------------
 *
 *  INITIALIZATION
 *
 *  -------------------------------------------------------------------------
 */

static struct cdevsw nit_cdev = {
	.d_name = "nit",
	.d_str = &nitinfo,
	.d_flag = D_CLONE | D_MP,
	.d_fop = &nit_ops,
	.d_mode = S_IFCHR | S_IRUGO | S_IWUGO,
	.d_kmod = THIS_MODULE,
};

/* 
 *  -------------------------------------------------------------------------
 *
 *  REGISTRATION
 *
 *  -------------------------------------------------------------------------
 */

streams_fastcall int
register_nit(struct cdevsw *cdev)
{
	int err;
	struct devnode *cmin, *node;

	err = -ENOMEM;
	if (!(cmin = kmalloc(2 * sizeof(*cmin), GFP_ATOMIC))) {
		goto error;
	}
	memset(cmin, 0, sizeof(*cmin));
	INIT_LIST_HEAD(&cmin->n_list);
	INIT_LIST_HEAD(&cmin->n_hash);
	cmin->n_name = cdev->d_name;
	cmin->n_str = cdev->d_str;
	cmin->n_flag = nit_cdev.d_flag;
	cmin->n_modid = cdev->d_modid;
	cmin->n_count = (atomic_t) ATOMIC_INIT(0);
	cmin->n_sqlvl = cdev->d_sqlvl;
	cmin->n_syncq = cdev->d_syncq;
	cmin->n_kmod = cdev->d_kmod;
	cmin->n_major = nit_cdev.d_major;
	cmin->n_mode = nit_cdev.d_mode;
	cmin->n_minor = cdev->d_major;
	cmin->n_dev = &nit_cdev;
	if ((err = register_strnod(&nit_cdev, cmin, cdev->d_major)) < 0) {
		kfree(cmin);
		goto error;
	}
	node = cmin + 1;
	memset(node, 0, sizeof(*node));
	INIT_LIST_HEAD(&node->n_list);
	INIT_LIST_HEAD(&node->n_hash);
	node->n_name = "nit";
	node->n_str = cdev->d_str;
	node->n_flag = D_CLONE;
	node->n_modid = cdev->d_modid;
	node->n_count = (atomic_t) ATOMIC_INIT(0);
	node->n_sqlvl = cdev->d_sqlvl;
	node->n_syncq = cdev->d_syncq;
	node->n_kmod = cdev->d_kmod;
	node->n_major = cdev->d_major;
	node->n_mode = cdev->d_mode;
	node->n_minor = NIT_NODE;
	node->n_dev = cdev;
	if ((err = register_strnod(cdev, node, NIT_NODE)) < 0) {
		unregister_strnod(&nit_cdev, cdev->d_major);
		kfree(cmin);
		goto error;
	}
      error:
	return (err);
}

EXPORT_SYMBOL_GPL(register_nit);

streams_fastcall int
unregister_nit(struct cdevsw *cdev)
{
	int err;
	struct devnode *cmin;

	err = -ENXIO;
	if (!(cmin = cmin_get(&nit_cdev, cdev->d_major)))
		goto error;
	if ((err = unregister_strnod(cdev, NIT_NODE)))
		goto error;
	if ((err = unregister_strnod(&nit_cdev, cdev->d_major)))
		goto error;
	kfree(cmin);
      error:
	return (err);
}

EXPORT_SYMBOL_GPL(unregister_nit);

/* 
 *  -------------------------------------------------------------------------
 *
 *  Special open for nit devices.
 *
 *  -------------------------------------------------------------------------
 */

/*
 *  nit_open: - open a nit device node
 *  @inode: the external filesystem inode
 *  @file: the external filesystem file pointer
 *
 *  nit_open() is only used to open a nit device from a character device node in an external
 *  filesystem.  This is never called for direct opens of a specfs device node (for direct opens see
 *  spec_dev_open() in strspecfs.c).  The character device number from the inode is used to
 *  determine the shadow special filesystem (internal) inode and chain the open call.
 *
 *  This is the separation point where we convert the external device number to an internal device
 *  number.  The external device number is contained in inode->i_rdev.
 */
STATIC int
nit_open(struct inode *inode, struct file *file)
{
	int err;
	struct cdevsw *cdev;
	major_t major;
	minor_t minor;
	modID_t modid, instance;

#if defined HAVE_KFUNC_TO_KDEV_T
	minor = MINOR(kdev_t_to_nr(inode->i_rdev));
	major = MAJOR(kdev_t_to_nr(inode->i_rdev));
#else
	minor = MINOR(inode->i_rdev);
	major = MAJOR(inode->i_rdev);
#endif
	minor = cdev_minor(&nit_cdev, major, minor);
	major = nit_cdev.d_major;
	modid = nit_cdev.d_modid;
	err = -ENXIO;
	if (!(cdev = sdev_get(minor))) {
		goto exit;
	}
	instance = cdev->d_modid;
	err = spec_open(file, cdev, makedevice(modid, instance), CLONEOPEN);
	sdev_put(cdev);
      exit:
	return (err);
}

STATIC struct file_operations nit_f_ops ____cacheline_aligned = {
	.owner = NULL,		/* yes NULL */
	.open = nit_open,
};

/* 
 *  -------------------------------------------------------------------------
 *
 *  INITIALIZATION
 *
 *  -------------------------------------------------------------------------
 */

#ifdef CONFIG_STREAMS_NIT_MODULE
static
#endif
int __init
nitinit(void)
{
	int err;

#ifdef CONFIG_STREAMS_NIT_MODULE
	printk(KERN_INFO NIT_BANNER);
#else
	printk(KERN_INFO NIT_SPLASH);
#endif
	nit_minfo.mi_idnum = nit_modid;
	if ((err = register_cmajor(&nit_cdev, major, &nit_f_ops)) < 0)
		return (err);
	if (major == 0 && err > 0)
		major = err;
	return (0);
};

#ifdef CONFIG_STREAMS_NIT_MODULE
static
#endif
void __exit
nitexit(void)
{
	if (unregister_cmajor(&nit_cdev, major) != 0)
		swerr();
};

#ifdef CONFIG_STREAMS_NIT_MODULE
module_init(nitinit);
module_exit(nitexit);
#endif
